### PR TITLE
[Pickles] Use the same 'permutation shifts' between different domains

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR is a small optimisation (split off from a larger fix to the wrap circuit's handling of domains). This builds upon #11607, although it is technically orthogonal.

This PR
* uses the same 'permutation shifts' (for the kimchi permutation argument) for all domains, regardless of their size
* adds an assertion to ensure that the shifts we use are consistent across all domains used in the step and wrap circuits.

This reduces the number of constraints used by approximately `14*num_inductive_rules`. For example, in the txn snark's wrap circuit, we save `14*5 = 70` constraints.

The original behaviour is preserved, guarded by a flag `disabled_not_the_same` hard-coded to `true`. This makes it trivial to restore the old behaviour if we later change to a larger domain that doesn't satisfy this property.

Checklist:

- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them

### CODE REVIEW

File: src/lib/pickles/pseudo/pseudo.ml [hunk 1]
```diff
@@ -43,12 +43,26 @@ module Make (Impl : Snarky_backendless.Snark_intf.Run) = struct
...
+          in
+          let disabled_not_the_same = true in
+          if all_the_same then Array.map ~f:Field.constant shifts
+          else if disabled_not_the_same then
+            failwith "Pseudo.Domain.shifts: found variable shifts"
+          else
+            let open Pickles_types.Plonk_types.Shifts in
+            let mk f = mask which (Vector.map all_shifts ~f) in
+            Array.init num_shifts ~f:(fun i ->
+                mk (fun a -> Field.constant a.(i)) )
```
Comment (mimoo) [inline src/lib/pickles/pseudo/pseudo.ml:65]: maybe clearer:

```ocaml
let const_shiftss = 
  let const_shifts x = Array.map x ~f:Field.constant in
  Vector.map all_shifts ~f:const_shifts 
in
let fold_ith_shifts i = 
  let consts = Vector.map all_shifts ~f:(fun x -> x.(i)) in
  mask which consts
in
Array.init num_shifts ~f:fold_ith_shifts
```

**Review by mimoo — APPROVED** [review]
LGTM

File: src/lib/pickles/pseudo/pseudo.ml [hunk 2]
```diff
@@ -43,12 +43,26 @@ module Make (Impl : Snarky_backendless.Snark_intf.Run) = struct
     let num_shifts = Nat.to_int Pickles_types.Plonk_types.Permuts.n
```
[the remark this answers is not in the document]
Comment (mrmr1993) [reply src/lib/pickles/pseudo/pseudo.ml:43]: Honestly, no idea.

File: src/lib/pickles/pseudo/pseudo.ml [hunk 3]
```diff
@@ -43,12 +43,26 @@ module Make (Impl : Snarky_backendless.Snark_intf.Run) = struct
...
+        ~(shifts : log2_size:int -> Field.Constant.t array) :
         Field.t Pickles_types.Plonk_types.Shifts.t =
-      let shifts = Vector.map log2s ~f:(fun d -> shifts ~log2_size:d) in
-      let open Pickles_types.Plonk_types.Shifts in
-      let mk f = mask which (Vector.map shifts ~f) in
-      Array.init num_shifts ~f:(fun i -> mk (fun a -> a.(i)))
+      let all_shifts = Vector.map log2s ~f:(fun d -> shifts ~log2_size:d) in
+      match all_shifts with
+      | [] ->
+          Array.init num_shifts ~f:(fun _ -> Field.zero)
```
[the remark this answers is not in the document]
Comment (mrmr1993) [reply src/lib/pickles/pseudo/pseudo.ml]: I've replaced it with a `failwith`; it was really just a placeholder value.

File: src/lib/pickles/pseudo/pseudo.ml [hunk 4]
```diff
@@ -43,12 +43,26 @@ module Make (Impl : Snarky_backendless.Snark_intf.Run) = struct
...
+      | shifts :: other_shiftss ->
+          let all_the_same =
+            Vector.for_all other_shiftss
+              ~f:(Array.for_all2_exn ~f:Field.Constant.equal shifts)
```
[the remark this answers is not in the document]
Comment (mrmr1993) [reply src/lib/pickles/pseudo/pseudo.ml:77]: Done, lmk if those comments give enough exposition.

File: src/lib/pickles/pseudo/pseudo.ml (hunk 2 above)
[the remark this answers is not in the document]
Comment (mimoo) [reply src/lib/pickles/pseudo/pseudo.ml:43]: @imeckler maybe :D?

[the remark this answers is not in the document]
Comment (imeckler) [reply src/lib/pickles/pseudo/pseudo.ml:43]: I don't know why, I just had to come up with a name. I thought like, a list of values of type `t` plus an index into that list is *like* a value of type `t`, so it's a "pseudo" `t`.

**Pull request merged**